
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Describing a simple experiment with isa objects &#8212; The ISA cookbook</title>
    
  <link href="../_static/css/theme.css" rel="stylesheet" />
  <link href="../_static/css/index.c5995385ac14fb8791e8eb36b4908be2.css" rel="stylesheet" />

    
  <link rel="stylesheet"
    href="../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/sphinx-book-theme.5f77b4aec8189eecf79907ce328c390d.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../_static/custom.css" />
    <link rel="stylesheet" type="text/css" href="../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../_static/js/index.1c5a1a01449ed65a7b51.js">

    <script id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/togglebutton.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script src="../_static/mermaid.js"></script>
    <script src="../_static/mermaid.min.js"></script>
    <script src="../_static/mermaid.core.js"></script>
    <script async="async" kind="hypothesis" src="https://hypothes.is/embed.js"></script>
    <script kind="utterances">

    var commentsRunWhenDOMLoaded = cb => {
    if (document.readyState != 'loading') {
        cb()
    } else if (document.addEventListener) {
        document.addEventListener('DOMContentLoaded', cb)
    } else {
        document.attachEvent('onreadystatechange', function() {
        if (document.readyState == 'complete') cb()
        })
    }
}

var addUtterances = () => {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src = "https://utteranc.es/client.js";
    script.async = "async";

    script.setAttribute("repo", "isa-tools/isa-api");
    script.setAttribute("issue-term", "pathname");
    script.setAttribute("theme", "github-light");
    script.setAttribute("label", "üí¨ comment");
    script.setAttribute("crossorigin", "anonymous");

    sections = document.querySelectorAll("div.section");
    if (sections !== null) {
        section = sections[sections.length-1];
        section.appendChild(script);
    }
}
commentsRunWhenDOMLoaded(addUtterances);
</script>
    <script >var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../_static/sphinx-book-theme.12a9622fbb08dcb3a2a40b2c02b83a57.js"></script>
    <script async="async" src="https://unpkg.com/thebelab@latest/lib/index.js"></script>
    <script >
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../_static/sphinx-thebe.js"></script>
    <link rel="canonical" href="https://isa-tools.github.io/isa-cookbook/content/creationtutorial.html" />
    <link rel="shortcut icon" href="../_static/isa.ico"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Validating ISA-Tab and ISA-JSON" href="validation.html" />
    <link rel="prev" title="Creating ISA content" href="creation.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="en" />
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../index.html">
      
      <img src="../_static/isa_footer_logo.svg" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">The ISA cookbook</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search for FAIR recipes..." aria-label="Search for FAIR recipes..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main navigation">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="index.html">
   ISA tools API
  </a>
 </li>
</ul>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="installation.html">
   ISA-API installation
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="isamodel.html">
   ISA data model
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="reading.html">
   Reading in ISA-Tab or ISA JSON
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="creation.html">
   Creating ISA content
  </a>
 </li>
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   Describing a simple experiment with
   <code class="docutils literal notranslate">
    <span class="pre">
     isa
    </span>
    <span class="pre">
     objects
    </span>
   </code>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="validation.html">
   Validating ISA-Tab and ISA-JSON
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="conversions.html">
   Converting between ISA format
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="importdata.html">
   Importing data into ISA formats
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="exportdata.html">
   Exporting data from ISA formats
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="examples.html">
   ISA API example programs
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="examples/example-createSimpleISAtab.html">
     createSimpleISAtab.py
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="examples/example-createSimpleISAJSON.html">
     createSimpleISAjson.py
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="examples/example-validateISAtab.html">
     validateISAtab.py
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="examples/example-validateISAjson.html">
     validateISAjson.py
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="github.html">
   Downloading files stored in Github
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="knownissues.html">
   Release Notes
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="notebooks.html">
   Hands-on Practicals
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/>
  <label for="toctree-checkbox-2">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="notebooks/reading-isa-tab-from-disk-exploring-and-validating.html">
     Reading ISA-Tab from files and Validating ISA-Tab files
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="notebooks/programmatic-rebuild-of-BII-S-3-with-ISA-API.html">
     Programmatic recreation of BII-S-3 with the ISA-API
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="notebooks/isa-protocol-type-synonym-effect-on-assay-table.html">
     ISA-API built-in semantic requirements for
     <code class="docutils literal notranslate">
      <span class="pre">
       protocol
      </span>
      <span class="pre">
       type
      </span>
     </code>
     values
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="notebooks/create-a-simple-ISA-descriptor.html">
     Create a simple ISA descriptor
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="notebooks/create_2x2x2_factorial_design.html">
     Create an ISA Study representing a 2x2x2 factorial design with single measurements
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="notebooks/create-a-repeated-treatment-design-with-ISA-descriptor.html">
     Create a repeated treatment design with ISA descriptor
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="notebooks/light-sensitity-study-repeated-treatment-design.html">
     Light Sensitivity Experiment:
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="notebooks/isa-metabolomics-study-design-create.html">
     ISA Create Mode example
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="notebooks/investigation-from-datascriptor-config-observational-variables-and-ontology-annotation.html">
     Create ISA-API Investigation from Datascriptor Study Design configuration
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="notebooks/isa-tab-unused-sources-not-serialized-brapi.html">
     Known Issue: unused ISA Source aren‚Äôt serialized to ISA-Tab
    </a>
   </li>
  </ul>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../_sources/content/creationtutorial.md"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.md</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        <a class="repository-button"
            href="https://github.com/isa-tools/isa-api"><button type="button" class="btn btn-secondary topbarbtn"
                data-toggle="tooltip" data-placement="left" title="Source repository"><i
                    class="fab fa-github"></i>repository</button></a>
        <a class="issues-button"
            href="https://github.com/isa-tools/isa-api/issues/new?title=Issue%20on%20page%20%2Fcontent/creationtutorial.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        <a class="edit-button" href="https://github.com/isa-tools/isa-api/edit/master/isa-cookbook/content/creationtutorial.md"><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Edit this page"><i class="fas fa-pencil-alt"></i>suggest edit</button></a>
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show">
            
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="section" id="describing-a-simple-experiment-with-isa-objects">
<h1>Describing a simple experiment with <code class="docutils literal notranslate"><span class="pre">isa</span> <span class="pre">objects</span></code><a class="headerlink" href="#describing-a-simple-experiment-with-isa-objects" title="Permalink to this headline">¬∂</a></h1>
<p>In this section, we provide a basic example of creating a complete
experiment descriptor using the ISA API‚Äôs model classes. The descriptor
is not complete and realistic, but it demonstrates the range of
component classes that you can use to create ISA content, including
things like sample characteristics, ontology annotations and units.</p>
<hr class="docutils" />
<div class="tip admonition">
<p class="admonition-title">IMPORTANT</p>
<p>As a pre-requisite to using ISA model please make sure you have read and
understood the <code class="docutils literal notranslate"><span class="pre">ISA</span> <span class="pre">Abstract</span> <span class="pre">Model</span> <span class="pre">&lt;/isamodel&gt;</span></code>{.interpreted-text
role=‚Äùdoc‚Äù} that the ISA formats are based on.</p>
</div>
<p>Firstly, we need to import the ISA API‚Äôs model classes from the
<code class="docutils literal notranslate"><span class="pre">isatools</span></code> PyPI package.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">isatools.model</span> <span class="kn">import</span> <span class="o">*</span>
</pre></div>
</div>
<p>Next, we build our descriptor encapsulated in a single Python function
to simplify the example code. In a real application or script, you might
decompose the functionality and hook it up to interactive components to
solicit feedback from a user on-the-fly.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">create_descriptor</span><span class="p">():</span>
    <span class="sd">&quot;&quot;&quot;Returns a simple but complete ISA-Tab 1.0 descriptor for illustration.&quot;&quot;&quot;</span>

    <span class="c1"># Create an empty Investigation object and set some values to the instance variables.</span>

    <span class="n">investigation</span> <span class="o">=</span> <span class="n">Investigation</span><span class="p">()</span>
    <span class="n">investigation</span><span class="o">.</span><span class="n">identifier</span> <span class="o">=</span> <span class="s2">&quot;i1&quot;</span>
    <span class="n">investigation</span><span class="o">.</span><span class="n">title</span> <span class="o">=</span> <span class="s2">&quot;My Simple ISA Investigation&quot;</span>
    <span class="n">investigation</span><span class="o">.</span><span class="n">description</span> <span class="o">=</span> <span class="s2">&quot;We could alternatively use the class constructor&#39;s parameters to set some default &quot;</span> \
                                <span class="s2">&quot;values at the time of creation, however we want to demonstrate how to use the &quot;</span> \
                                <span class="s2">&quot;object&#39;s instance variables to set values.&quot;</span>
    <span class="n">investigation</span><span class="o">.</span><span class="n">submission_date</span> <span class="o">=</span> <span class="s2">&quot;2016-11-03&quot;</span>
    <span class="n">investigation</span><span class="o">.</span><span class="n">public_release_date</span> <span class="o">=</span> <span class="s2">&quot;2016-11-03&quot;</span>

    <span class="c1"># Create an empty Study object and set some values. The Study must have a filename, otherwise when we serialize it</span>
    <span class="c1"># to ISA-Tab we would not know where to write it. We must also attach the study to the investigation by adding it</span>
    <span class="c1"># to the &#39;investigation&#39; object&#39;s list of studies.</span>

    <span class="n">study</span> <span class="o">=</span> <span class="n">Study</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="s2">&quot;s_study.txt&quot;</span><span class="p">)</span>
    <span class="n">study</span><span class="o">.</span><span class="n">identifier</span> <span class="o">=</span> <span class="s2">&quot;s1&quot;</span>
    <span class="n">study</span><span class="o">.</span><span class="n">title</span> <span class="o">=</span> <span class="s2">&quot;My ISA Study&quot;</span>
    <span class="n">study</span><span class="o">.</span><span class="n">description</span> <span class="o">=</span> <span class="s2">&quot;Like with the Investigation, we could use the class constructor to set some default values, &quot;</span> \
                        <span class="s2">&quot;but have chosen to demonstrate in this example the use of instance variables to set initial &quot;</span> \
                        <span class="s2">&quot;values.&quot;</span>
    <span class="n">study</span><span class="o">.</span><span class="n">submission_date</span> <span class="o">=</span> <span class="s2">&quot;2016-11-03&quot;</span>
    <span class="n">study</span><span class="o">.</span><span class="n">public_release_date</span> <span class="o">=</span> <span class="s2">&quot;2016-11-03&quot;</span>
    <span class="n">investigation</span><span class="o">.</span><span class="n">studies</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">study</span><span class="p">)</span>

    <span class="c1"># Some instance variables are typed with different objects and lists of objects. For example, a Study can have a</span>
    <span class="c1"># list of design descriptors. A design descriptor is an Ontology Annotation describing the kind of study at hand.</span>
    <span class="c1"># Ontology Annotations should typically reference an Ontology Source. We demonstrate a mix of using the class</span>
    <span class="c1"># constructors and setting values with instance variables. Note that the OntologyAnnotation object</span>
    <span class="c1"># &#39;intervention_design&#39; links its &#39;term_source&#39; directly to the &#39;obi&#39; object instance. To ensure the OntologySource</span>
    <span class="c1"># is encapsulated in the descriptor, it is added to a list of &#39;ontology_source_references&#39; in the Investigation</span>
    <span class="c1"># object. The &#39;intervention_design&#39; object is then added to the list of &#39;design_descriptors&#39; held by the Study</span>
    <span class="c1"># object.</span>

    <span class="n">obi</span> <span class="o">=</span> <span class="n">OntologySource</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s1">&#39;OBI&#39;</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="s2">&quot;Ontology for Biomedical Investigations&quot;</span><span class="p">)</span>
    <span class="n">investigation</span><span class="o">.</span><span class="n">ontology_source_references</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">obi</span><span class="p">)</span>
    <span class="n">intervention_design</span> <span class="o">=</span> <span class="n">OntologyAnnotation</span><span class="p">(</span><span class="n">term_source</span><span class="o">=</span><span class="n">obi</span><span class="p">)</span>
    <span class="n">intervention_design</span><span class="o">.</span><span class="n">term</span> <span class="o">=</span> <span class="s2">&quot;intervention design&quot;</span>
    <span class="n">intervention_design</span><span class="o">.</span><span class="n">term_accession</span> <span class="o">=</span> <span class="s2">&quot;http://purl.obolibrary.org/obo/OBI_0000115&quot;</span>
    <span class="n">study</span><span class="o">.</span><span class="n">design_descriptors</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">intervention_design</span><span class="p">)</span>

    <span class="c1"># Other instance variables common to both Investigation and Study objects include &#39;contacts&#39; and &#39;publications&#39;,</span>
    <span class="c1"># each with lists of corresponding Person and Publication objects.</span>

    <span class="n">contact</span> <span class="o">=</span> <span class="n">Person</span><span class="p">(</span><span class="n">first_name</span><span class="o">=</span><span class="s2">&quot;Alice&quot;</span><span class="p">,</span> <span class="n">last_name</span><span class="o">=</span><span class="s2">&quot;Robertson&quot;</span><span class="p">,</span> <span class="n">affiliation</span><span class="o">=</span><span class="s2">&quot;University of Life&quot;</span><span class="p">,</span> <span class="n">roles</span><span class="o">=</span><span class="p">[</span><span class="n">OntologyAnnotation</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s1">&#39;submitter&#39;</span><span class="p">)])</span>
    <span class="n">study</span><span class="o">.</span><span class="n">contacts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">contact</span><span class="p">)</span>
    <span class="n">publication</span> <span class="o">=</span> <span class="n">Publication</span><span class="p">(</span><span class="n">title</span><span class="o">=</span><span class="s2">&quot;Experiments with Elephants&quot;</span><span class="p">,</span> <span class="n">author_list</span><span class="o">=</span><span class="s2">&quot;A. Robertson, B. Robertson&quot;</span><span class="p">)</span>
    <span class="n">publication</span><span class="o">.</span><span class="n">pubmed_id</span> <span class="o">=</span> <span class="s2">&quot;12345678&quot;</span>
    <span class="n">publication</span><span class="o">.</span><span class="n">status</span> <span class="o">=</span> <span class="n">OntologyAnnotation</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s2">&quot;published&quot;</span><span class="p">)</span>
    <span class="n">study</span><span class="o">.</span><span class="n">publications</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">publication</span><span class="p">)</span>

    <span class="c1"># To create the study graph that corresponds to the contents of the study table file (the s_*.txt file), we need</span>
    <span class="c1"># to create a process sequence. To do this we use the Process class and attach it to the Study object&#39;s</span>
    <span class="c1"># &#39;process_sequence&#39; list instance variable. Each process must be linked with a Protocol object that is attached to</span>
    <span class="c1"># a Study object&#39;s &#39;protocols&#39; list instance variable. The sample collection Process object usually has as input</span>
    <span class="c1"># a Source material and as output a Sample material.</span>

    <span class="c1"># Here we create one Source material object and attach it to our study.</span>

    <span class="n">source</span> <span class="o">=</span> <span class="n">Source</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s1">&#39;source_material&#39;</span><span class="p">)</span>
    <span class="n">study</span><span class="o">.</span><span class="n">sources</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">source</span><span class="p">)</span>

    <span class="c1"># Then we create three Sample objects, with organism as Homo Sapiens, and attach them to the study. We use the utility function</span>
    <span class="c1"># batch_create_material() to clone a prototype material object. The function automatiaclly appends</span>
    <span class="c1"># an index to the material name. In this case, three samples will be created, with the names</span>
    <span class="c1"># &#39;sample_material-0&#39;, &#39;sample_material-1&#39; and &#39;sample_material-2&#39;.</span>

    <span class="n">prototype_sample</span> <span class="o">=</span> <span class="n">Sample</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s1">&#39;sample_material&#39;</span><span class="p">,</span> <span class="n">derives_from</span><span class="o">=</span><span class="p">[</span><span class="n">source</span><span class="p">])</span>

    <span class="n">ncbitaxon</span> <span class="o">=</span> <span class="n">OntologySource</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s1">&#39;NCBITaxon&#39;</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="s2">&quot;NCBI Taxonomy&quot;</span><span class="p">)</span>
    <span class="n">investigation</span><span class="o">.</span><span class="n">ontology_source_references</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ncbitaxon</span><span class="p">)</span> <span class="c1"># remember to add the newly declared ontology source to the parent investigation</span>

    <span class="n">characteristic_organism</span> <span class="o">=</span> <span class="n">Characteristic</span><span class="p">(</span><span class="n">category</span><span class="o">=</span><span class="n">OntologyAnnotation</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s2">&quot;Organism&quot;</span><span class="p">),</span>
                                     <span class="n">value</span><span class="o">=</span><span class="n">OntologyAnnotation</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s2">&quot;Homo Sapiens&quot;</span><span class="p">,</span> <span class="n">term_source</span><span class="o">=</span><span class="n">ncbitaxon</span><span class="p">,</span>
                                                              <span class="n">term_accession</span><span class="o">=</span><span class="s2">&quot;http://purl.bioontology.org/ontology/NCBITAXON/9606&quot;</span><span class="p">))</span>
    <span class="n">prototype_sample</span><span class="o">.</span><span class="n">characteristics</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">characteristic_organism</span><span class="p">)</span>

    <span class="n">study</span><span class="o">.</span><span class="n">samples</span> <span class="o">=</span> <span class="n">batch_create_materials</span><span class="p">(</span><span class="n">prototype_sample</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>  <span class="c1"># creates a batch of 3 samples</span>

    <span class="c1"># Now we create a single Protocol object that represents our sample collection protocol, and attach it to the</span>
    <span class="c1"># study object. Protocols must be declared before we describe Processes, as a processing event of some sort</span>
    <span class="c1"># must execute some defined protocol. In the case of the class model, Protocols should therefore be declared</span>
    <span class="c1"># before Processes in order for the Process to be linked to one.</span>

    <span class="n">sample_collection_protocol</span> <span class="o">=</span> <span class="n">Protocol</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;sample collection&quot;</span><span class="p">,</span>
                                          <span class="n">protocol_type</span><span class="o">=</span><span class="n">OntologyAnnotation</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s2">&quot;sample collection&quot;</span><span class="p">))</span>
    <span class="n">study</span><span class="o">.</span><span class="n">protocols</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">sample_collection_protocol</span><span class="p">)</span>
    <span class="n">sample_collection_process</span> <span class="o">=</span> <span class="n">Process</span><span class="p">(</span><span class="n">executes_protocol</span><span class="o">=</span><span class="n">sample_collection_protocol</span><span class="p">)</span>

    <span class="c1"># Next, we link our materials to the Process. In this particular case, we are describing a sample collection</span>
    <span class="c1"># process that takes one source material, and produces three different samples.</span>
    <span class="c1">#</span>
    <span class="c1"># (source_material)-&gt;(sample collection)-&gt;[(sample_material-0), (sample_material-1), (sample_material-2)]</span>

    <span class="k">for</span> <span class="n">src</span> <span class="ow">in</span> <span class="n">study</span><span class="o">.</span><span class="n">sources</span><span class="p">:</span>
        <span class="n">sample_collection_process</span><span class="o">.</span><span class="n">inputs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">src</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">sam</span> <span class="ow">in</span> <span class="n">study</span><span class="o">.</span><span class="n">samples</span><span class="p">:</span>
        <span class="n">sample_collection_process</span><span class="o">.</span><span class="n">outputs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">sam</span><span class="p">)</span>

    <span class="c1"># Finally, attach the finished Process object to the study process_sequence. This can be done many times to</span>
    <span class="c1"># describe multiple sample collection events.</span>

    <span class="n">study</span><span class="o">.</span><span class="n">process_sequence</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">sample_collection_process</span><span class="p">)</span>

    <span class="c1"># Next, we build n Assay object and attach two protocols, extraction and sequencing.</span>

    <span class="n">assay</span> <span class="o">=</span> <span class="n">Assay</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="s2">&quot;a_assay.txt&quot;</span><span class="p">)</span>
    <span class="n">extraction_protocol</span> <span class="o">=</span> <span class="n">Protocol</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s1">&#39;extraction&#39;</span><span class="p">,</span> <span class="n">protocol_type</span><span class="o">=</span><span class="n">OntologyAnnotation</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s2">&quot;material extraction&quot;</span><span class="p">))</span>
    <span class="n">study</span><span class="o">.</span><span class="n">protocols</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">extraction_protocol</span><span class="p">)</span>
    <span class="n">sequencing_protocol</span> <span class="o">=</span> <span class="n">Protocol</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s1">&#39;sequencing&#39;</span><span class="p">,</span> <span class="n">protocol_type</span><span class="o">=</span><span class="n">OntologyAnnotation</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s2">&quot;material sequencing&quot;</span><span class="p">))</span>
    <span class="n">study</span><span class="o">.</span><span class="n">protocols</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">sequencing_protocol</span><span class="p">)</span>

    <span class="c1"># To build out assay graphs, we enumereate the samples from the study-level, and for each sample we create an</span>
    <span class="c1"># extraction process and a sequencing process. The extraction process takes as input a sample material, and produces</span>
    <span class="c1"># an extract material. The sequencing process takes the extract material and produces a data file. This will</span>
    <span class="c1"># produce three graphs, from sample material through to data, as follows:</span>
    <span class="c1">#</span>
    <span class="c1"># (sample_material-0)-&gt;(extraction)-&gt;(extract-0)-&gt;(sequencing)-&gt;(sequenced-data-0)</span>
    <span class="c1"># (sample_material-1)-&gt;(extraction)-&gt;(extract-1)-&gt;(sequencing)-&gt;(sequenced-data-1)</span>
    <span class="c1"># (sample_material-2)-&gt;(extraction)-&gt;(extract-2)-&gt;(sequencing)-&gt;(sequenced-data-2)</span>
    <span class="c1">#</span>
    <span class="c1"># Note that the extraction processes and sequencing processes are distinctly separate instances, where the three</span>
    <span class="c1"># graphs are NOT interconnected.</span>

    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">sample</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">study</span><span class="o">.</span><span class="n">samples</span><span class="p">):</span>

        <span class="c1"># create an extraction process that executes the extraction protocol</span>

        <span class="n">extraction_process</span> <span class="o">=</span> <span class="n">Process</span><span class="p">(</span><span class="n">executes_protocol</span><span class="o">=</span><span class="n">extraction_protocol</span><span class="p">)</span>

        <span class="c1"># extraction process takes as input a sample, and produces an extract material as output</span>

        <span class="n">extraction_process</span><span class="o">.</span><span class="n">inputs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">sample</span><span class="p">)</span>
        <span class="n">material</span> <span class="o">=</span> <span class="n">Material</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;extract-</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">))</span>
        <span class="n">material</span><span class="o">.</span><span class="n">type</span> <span class="o">=</span> <span class="s2">&quot;Extract Name&quot;</span>
        <span class="n">extraction_process</span><span class="o">.</span><span class="n">outputs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">material</span><span class="p">)</span>

        <span class="c1"># create a sequencing process that executes the sequencing protocol</span>

        <span class="n">sequencing_process</span> <span class="o">=</span> <span class="n">Process</span><span class="p">(</span><span class="n">executes_protocol</span><span class="o">=</span><span class="n">sequencing_protocol</span><span class="p">)</span>
        <span class="n">sequencing_process</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="s2">&quot;assay-name-</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
        <span class="n">sequencing_process</span><span class="o">.</span><span class="n">inputs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">extraction_process</span><span class="o">.</span><span class="n">outputs</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>

        <span class="c1"># Sequencing process usually has an output data file</span>

        <span class="n">datafile</span> <span class="o">=</span> <span class="n">DataFile</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="s2">&quot;sequenced-data-</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">),</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;Raw Data File&quot;</span><span class="p">)</span>
        <span class="n">sequencing_process</span><span class="o">.</span><span class="n">outputs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">datafile</span><span class="p">)</span>

        <span class="c1"># Ensure Processes are linked forward and backward. plink(from_process, to_process) is a function to set</span>
        <span class="c1"># these links for you. It is found in the isatools.model package</span>

        <span class="n">plink</span><span class="p">(</span><span class="n">extraction_process</span><span class="p">,</span> <span class="n">sequencing_process</span><span class="p">)</span>

        <span class="c1"># make sure the extract, data file, and the processes are attached to the assay</span>

        <span class="n">assay</span><span class="o">.</span><span class="n">data_files</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">datafile</span><span class="p">)</span>
        <span class="n">assay</span><span class="o">.</span><span class="n">samples</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">sample</span><span class="p">)</span>
        <span class="n">assay</span><span class="o">.</span><span class="n">other_material</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">material</span><span class="p">)</span>
        <span class="n">assay</span><span class="o">.</span><span class="n">process_sequence</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">extraction_process</span><span class="p">)</span>
        <span class="n">assay</span><span class="o">.</span><span class="n">process_sequence</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">sequencing_process</span><span class="p">)</span>
        <span class="n">assay</span><span class="o">.</span><span class="n">measurement_type</span> <span class="o">=</span> <span class="n">OntologyAnnotation</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s2">&quot;gene sequencing&quot;</span><span class="p">)</span>
        <span class="n">assay</span><span class="o">.</span><span class="n">technology_type</span> <span class="o">=</span> <span class="n">OntologyAnnotation</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s2">&quot;nucleotide sequencing&quot;</span><span class="p">)</span>

    <span class="c1"># attach the assay to the study</span>

    <span class="n">study</span><span class="o">.</span><span class="n">assays</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">assay</span><span class="p">)</span>

    <span class="c1">#IMPORTANT: remember to list all Characteristics used in the study object: do as follows:</span>
    <span class="n">study</span><span class="o">.</span><span class="n">characteristic_categories</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">characteristic_organism</span><span class="o">.</span><span class="n">category</span><span class="p">)</span>
</pre></div>
</div>
<p>To write out the ISA-Tab, you can use the <code class="docutils literal notranslate"><span class="pre">isatab.dumps()</span></code> function:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">isatools</span> <span class="kn">import</span> <span class="n">isatab</span>
<span class="k">return</span> <span class="n">isatab</span><span class="o">.</span><span class="n">dumps</span><span class="p">(</span><span class="n">investigation</span><span class="p">)</span>  <span class="c1"># dumps() writes out the ISA as a string representation of the ISA-Tab</span>
</pre></div>
</div>
<p>The function listed above is designed to return all three files as a
single string output for ease of inspection. Alternatively you could do
something like <code class="docutils literal notranslate"><span class="pre">dump(isa_obj=investigation,</span> <span class="pre">output_path='./')</span></code> to write
the files to the file system.</p>
<p>Alternatively to write out the ISA JSON, you can use the
<code class="docutils literal notranslate"><span class="pre">ISAJSONEncoder</span></code> class with the Python <code class="docutils literal notranslate"><span class="pre">json</span></code> package:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">json</span>
<span class="kn">from</span> <span class="nn">isatools.isajson</span> <span class="kn">import</span> <span class="n">ISAJSONEncoder</span>
<span class="c1"># Note that the extra parameters sort_keys, indent and separators are to make the output more human-readable.</span>
<span class="k">return</span> <span class="n">json</span><span class="o">.</span><span class="n">dumps</span><span class="p">(</span><span class="n">investigation</span><span class="p">,</span> <span class="bp">cls</span><span class="o">=</span><span class="n">ISAJSONEncoder</span><span class="p">,</span> <span class="n">sort_keys</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">indent</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span> <span class="n">separators</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="p">,</span> <span class="s1">&#39;: &#39;</span><span class="p">))</span>
</pre></div>
</div>
<p>The final lines of code is a <code class="docutils literal notranslate"><span class="pre">main</span></code> routine to invoke the
<code class="docutils literal notranslate"><span class="pre">create_descriptor()</span></code> function.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s1">&#39;__main__&#39;</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">create_descriptor</span><span class="p">())</span>
</pre></div>
</div>
<p>If you save the file into something like <code class="docutils literal notranslate"><span class="pre">createSimpleISA.py</span></code>, to
execute this script on the command line and view the output, you would
run it with:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>python createSimpleISA.py
</pre></div>
</div>
<p>This example can be found in the <code class="docutils literal notranslate"><span class="pre">isatools.examples</span></code> package in
<a class="reference external" href="https://github.com/ISA-tools/isa-api/blob/master/isatools/examples/createSimpleISAtab.py">createSimpleISAtab.py</a>
and
<a class="reference external" href="https://github.com/ISA-tools/isa-api/blob/master/isatools/examples/createSimpleISAJSON.py">createSimpleISAJSON.py</a></p>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./content"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
        
        <div class='prev-next-bottom'>
            
    <a class='left-prev' id="prev-link" href="creation.html" title="previous page">Creating ISA content</a>
    <a class='right-next' id="next-link" href="validation.html" title="next page">Validating ISA-Tab and ISA-JSON</a>

        </div>
        
        </div>
    </div>
    <footer class="footer mt-5 mt-md-0">
    <div class="container">
      <p>
        
          By The ISA Team<br/>
        
            &copy; Copyright 2021 (CC-BY 4.0).<br/>
          <div class="extra_footer">
            <script>mermaid.init();</script> <div class="row_footer"> This page was created thanks to work by <a href="https://github.com/jupyter/jupyter-book/graphs/contributors">The Jupyter Book Community</a></div> <div class="row_footer"> This project is supported financially by the <a href="https://wellcome.org/">Welcome Trust</a>, Elixir-UK and the European Union</div>
<div class="row_footer"> <div class="column" style="center"><a href="https://wellcome.org/"><img src="../_static/images/logo/wellcome-logo-black.png" height="25" alt="Wellcome" style="float:center""></a></div> <div class="column"><a href="https://www.elixir-uk.org.eu/"><img src="../_static/images/logo/elixir-uk_transparent.png" height="40" alt="IMI logo" style="float:center""></a></div> <div class="column"><a href="https://https://ec.europa.eu/programmes/horizon2020/en"><img src="../_static/images/logo/eu-flag-recolored.png" height="25" alt="EFPIA logo" style="float:center""></a></div> </div>
          </div>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>
  
  <script src="../_static/js/index.1c5a1a01449ed65a7b51.js"></script>

  
  </body>
</html>